.github/workflows/pullrequest.yml

```diff
@@ -2,67 +2,144 @@ name: Pull-Request CI
 
 on:
   pull_request:
-    branches: [ release, develop ]
+    branches: [release, develop]
 
 jobs:
-  build_test:
-
+  unit-test:
+    name: Unit Tests
     runs-on: ubuntu-latest
-
     steps:
-      - uses: actions/checkout@v2
-      - name: Cache multiple paths
-        uses: actions/cache@v2
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      - name: Setup Node.js and cache
+        uses: actions/setup-node@v2
         with:
-          path: |
-            **/node_modules
-          key: ${{ runner.os }}-${{ hashFiles('package.json') }}
-      - name: Use Node.js 12.16.2
-        uses: actions/setup-node@v1
+          node-version-file: .nvmrc
+          cache: npm
+      - name: Setup Angular cache
+        uses: actions/cache@v2
         with:
-          node-version: 12.16.2
+          path: node_modules/.cache
+          key: pr-angular-${{ runner.os }}-${{ hashFiles('**/package-lock.json') }}
       - run: npm install --legacy-peer-deps
       - run: npm run build:prepare
       - run: npm run test:ci
-      - name: Run Cypress tests 🧪
+  e2e-tests:
+    name: E2E Tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      - name: Setup Node.js
+        uses: actions/setup-node@v2
+        with:
+          node-version-file: .nvmrc
+      - name: Run Cypress tests
         uses: cypress-io/github-action@v2
         with:
           command: npm run test:cypress
 
-  build_executables:
+  headless:
+    name: Build Headless
     runs-on: ${{ matrix.os }}
-
     strategy:
       matrix:
-        os: [ macos-latest, windows-latest, ubuntu-latest ]
-
+        os: [macos-latest, ubuntu-latest, windows-latest]
     steps:
-      - uses: actions/checkout@v2
-      - name: Cache multiple paths
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      - name: Setup Node.js
+        uses: actions/setup-node@v2
+        with:
+          node-version: 12.16.2
+          cache: npm
+      - name: Setup Angular cache
         uses: actions/cache@v2
         with:
-          path: |
-            **/node_modules
-          key: ${{ runner.os }}-${{ hashFiles('package.json') }}
-      - name: Use Node.js 12.16.2
-        uses: actions/setup-node@v1
+          path: node_modules/.cache
+          key: pr-angular-${{ runner.os }}-${{ hashFiles('**/package-lock.json') }}
+      - run: npm install
```

Comment:
the reason I had it all under one build group is that I didn't wanted to waste the time for
- `npm install`
- `npm build:prepare` 

each time, which installs it all, compiles angular stuff, compiles the basic server files and thats between 6-10 minutes

there are ways to share the output BUT I need try that again --- I think I had it done for the release pipeline

Reply:
I hoped splitting them up would speed up things. But the majority of time is spent in those two steps you mentioned.

Will move those back together again 👍 

What probably works is, to run `install` and `build:prepare`, then upload `node_modules` and the needed directories as artifact. In the following jobs, download said artifact into the current folder before anything else.

Reply:
yeah the node_modules probably not helping in terms of speed - those could stay in each build for now

but everything else - the built angular app and the built server output that can be shared betweet them all

also the `prepare:` don't need OS dependency - the created files are js nothing native there